### PR TITLE
Support `norm='forward'`/`'backward'` in `cupy.fft` functions

### LINKED ISSUE #944: Unnormalized inverse FFT
Looks like even with norm=None inverse fft normalizes the output, can we add a way to avoid it? significantly cripples efficiency.
https://github.com/cupy/cupy/blob/master/cupy/fft/fft.py#L84

### PR DESCRIPTION
**Pull request by leofang**
Take over #3318 with permission from @mattvend.

Close #944. Close #3318.

This PR follows the upstream change included since NumPy 1.20 (numpy/numpy#16476).

### CODE REVIEW

**Comment by leofang** [comment]
Question: I am not sure I am using `@pytest.fixture` right --- Is there a way to define just one fixture and make it seen by all tests in the same module? It'd be great to do so to avoid code dup!

**Comment by leofang** [comment]
Another question is I have to duplicate many test cases in `@testing.parametrize` just to test `norm="forward"`. Is this desired, or is there a way to further parametrize this?

**Comment by leofang** [comment]
> Question: I am not sure I am using `@pytest.fixture` right --- Is there a way to define just one fixture and make it seen by all tests in the same module? It'd be great to do so to avoid code dup!

Maybe `@pytest.mark.usefixtures`?

**Comment by leofang** [comment]
> Another question is I have to duplicate many test cases in `@testing.parametrize` just to test `norm="forward"`. Is this desired, or is there a way to further parametrize this?

I used `@testing.product_dict` to achieve this following some ndimage test usages. This should be ready now.

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
Jenkins, test this please

File: cupy/fft/_fft.py [hunk 1]
```diff
@@ -189,11 +189,14 @@ def _exec_fft(a, direction, value_type, norm, axis, overwrite_x,
     sz = out.shape[-1]
     if fft_type == cufft.CUFFT_R2C or fft_type == cufft.CUFFT_D2Z:
         sz = n
-    if norm is None:
+    if norm == 'backward':
         if direction == cufft.CUFFT_INVERSE:
             out /= sz
```
Comment (grlee77) [inline cupy/fft/_fft.py:194]:
```suggestion
    if norm == 'backward' and direction == cufft.CUFFT_INVERSE:
        out /= sz
```

File: cupy/fft/_fft.py [hunk 2]
```diff
@@ -189,11 +189,14 @@ def _exec_fft(a, direction, value_type, norm, axis, overwrite_x,
...
-    else:
+    elif norm == 'ortho':
         out /= math.sqrt(sz)
+    elif norm == 'forward':
+        if direction == cufft.CUFFT_FORWARD:
+            out /= sz
```
Comment (grlee77) [inline cupy/fft/_fft.py:199]:
```suggestion
    elif norm == 'forward' and direction == cufft.CUFFT_FORWARD:
        out /= sz
```

File: cupy/fft/_fft.py [hunk 3]
```diff
@@ -553,11 +557,14 @@ def _exec_fftn(a, direction, value_type, norm, axes, overwrite_x,
     # normalize by the product of the shape along the transformed axes
     arr = a if fft_type in (cufft.CUFFT_R2C, cufft.CUFFT_D2Z) else out
     sz = _prod([arr.shape[ax] for ax in axes])
-    if norm is None:
+    if norm == 'backward':
         if direction == cufft.CUFFT_INVERSE:
             out /= sz
```
Comment (grlee77) [inline cupy/fft/_fft.py:562]:
```suggestion
    if norm == 'backward' and direction == cufft.CUFFT_INVERSE:
        out /= sz
```

File: cupy/fft/_fft.py [hunk 4]
```diff
@@ -553,11 +557,14 @@ def _exec_fftn(a, direction, value_type, norm, axes, overwrite_x,
...
-    else:
+    elif norm == 'ortho':
         out /= math.sqrt(sz)
+    elif norm == 'forward':
+        if direction == cufft.CUFFT_FORWARD:
+            out /= sz
```
Comment (grlee77) [inline cupy/fft/_fft.py:567]:
```suggestion
    elif norm == 'forward' and direction == cufft.CUFFT_FORWARD:
        out /= sz
```

**Review by grlee77 — APPROVED** [review]
This looks great, thanks @leofang.

I made one (optional) style suggestion

**Comment by grlee77** [comment]
There are a few flake8 complaints related to f-strings

**Comment by leofang** [comment]
> There are a few flake8 complaints related to f-strings

Thanks, @grlee77! The flake8 errors are fixed in #4803, so it should now work after committing the changes you suggested 🙂

**Comment by leofang** [comment]
Jenkins, test this please

File: tests/cupy_tests/fft_tests/test_fft.py [hunk 5]
```diff
@@ -359,31 +368,36 @@ def test_fft_allocate(self):
...
+@pytest.mark.usefixtures('skip_forward_backward')
+@testing.parameterize(*(
+    testing.product_dict([
+        {'shape': (3, 4), 's': None, 'axes': None},
+        {'shape': (3, 4), 's': (1, None), 'axes': None},
+        {'shape': (3, 4), 's': (1, 5), 'axes': None},
+        {'shape': (3, 4), 's': None, 'axes': (-2, -1)},
+        {'shape': (3, 4), 's': None, 'axes': (-1, -2)},
+        {'shape': (3, 4), 's': None, 'axes': (0,)},
+        {'shape': (3, 4), 's': None, 'axes': None},
```
Comment (toslunar) [inline tests/cupy_tests/fft_tests/test_fft.py:380]: The triple (shape, s, axes) is repeated.

By the way, testing all `norm`s for each triple may be heavy.  Could you compare the numbers/time of tests between before and after the PR?

**Review by toslunar — CHANGES_REQUESTED** [review]
I reviewed the tests first.

Could you test an invalid `norm` value for each function?  I found `norm=''` (with `accept_error=ValueError`) is such a testcase, but not the all functions are covered.

File: tests/cupy_tests/fft_tests/test_fft.py (hunk 5 above)
Comment (leofang) [reply tests/cupy_tests/fft_tests/test_fft.py:380]: Sorry, you meant repeated with different `norm`s, right?

I checked that the number of tests is doubled (2 `norm`s -> 4 `norm`s, if testing `norm=''` then 5), but the increased time is negligible as FFT is very fast.

Comment (leofang) [reply tests/cupy_tests/fft_tests/test_fft.py:380]: @toslunar I found a weird NumPy behavior which I think is a bug that should be reported and fixed upstream:
```python
>>> import numpy as np
>>>
>>> a = np.random.random((2,3,4))
>>> b = np.fft.fft2(a, norm='')  # <-- this does not work (norm invalid)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<__array_function__ internals>", line 6, in fft2
  File "/home/leofang/miniconda3/envs/cupy_cuda112_dev/lib/python3.7/site-packages/numpy/fft/_pocketfft.py", line 1012, in fft2
    return _raw_fftnd(a, s, axes, fft, norm)
  File "/home/leofang/miniconda3/envs/cupy_cuda112_dev/lib/python3.7/site-packages/numpy/fft/_pocketfft.py", line 705, in _raw_fftnd
    a = function(a, n=s[ii], axis=axes[ii], norm=norm)
  File "<__array_function__ internals>", line 6, in fft
  File "/home/leofang/miniconda3/envs/cupy_cuda112_dev/lib/python3.7/site-packages/numpy/fft/_pocketfft.py", line 214, in fft
    inv_norm = _get_forward_norm(n, norm)
  File "/home/leofang/miniconda3/envs/cupy_cuda112_dev/lib/python3.7/site-packages/numpy/fft/_pocketfft.py", line 88, in _get_forward_norm
    raise ValueError(f'Invalid norm value {norm}; should be "backward",'
ValueError: Invalid norm value ; should be "backward","ortho" or "forward".
>>> b = np.fft.fft2(a, norm='', axes=())  # <-- this works (despite norm is invalid)?!
>>> (b == a).all()
True
```
If `axes=()` then `norm` is not checked (NumPy just returns the input array). Should I just skip testing this corner case?

Comment (leofang) [reply tests/cupy_tests/fft_tests/test_fft.py:380]: > If `axes=()` then `norm` is not checked (NumPy just returns the input array). Should I just skip testing this corner case?

Workaround done in f8bd913.

**Comment by leofang** [comment]
Jenkins, test this please

**Review by toslunar — COMMENTED** [review]
The docs and tests LGTM.

File: cupy/fft/_fft.py [hunk 6]
```diff
@@ -209,15 +210,10 @@ def _fft_c2c(a, direction, norm, axes, overwrite_x, plan=None):
 
 def _fft(a, s, axes, norm, direction, value_type='C2C', overwrite_x=False,
          plan=None):
-    if isinstance(a, np.ndarray):
+    if not isinstance(a, cupy.ndarray):
```
[the remark this answers is not in the document]
Comment (leofang) [reply cupy/fft/_fft.py:213]: It never accepted foreign arrays that support the CAI protocol, nor did we write a test for that need.
```python
Traceback (most recent call last):
  File "test_fft.py", line 28, in <module>
    dummy_b = cp.fft.fftn(dummy)
  File "/home/leofang/cupy_cuda110/cupy/fft/_fft.py", line 754, in fftn
    func = _default_fft_func(a, s, axes)
  File "/home/leofang/cupy_cuda110/cupy/fft/_fft.py", line 625, in _default_fft_func
    a.ndim == 1 or not config.enable_nd_planning):
AttributeError: 'DummyObjectWithCudaArrayInterface' object has no attribute 'ndim'
```
In fact my gut feeling is most of CuPy APIs do not directly support CAI. 

The check here was added mainly to prevent numpy arrays from being passed to high-level APIs when a multi-GPU transform is requested.

**Review by toslunar — APPROVED** [review]

**Comment by leofang** [comment]
@mattvend Thanks a lot for your contribution on this! It just got in.
